File: src/unittest/0250-file_ostream.py
#!/usr/bin/python
# -*- coding: utf-8; mode: python; tab-width: 3 -*-
#---------------------------------------------------------------------------------------------------
# Application-Building Components
# Copyright 2011-2013 Raffaello D. Di Napoli
#---------------------------------------------------------------------------------------------------
# This file is part of Application-Building Components (henceforth referred to as ABC).
#
# ABC is free software: you can redistribute it and/or modify it under the terms of the GNU General
# Public License as published by the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# ABC is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the
# implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
# Public License for more details.
#
# You should have received a copy of the GNU General Public License along with ABC. If not, see
# <http://www.gnu.org/licenses/>.
#---------------------------------------------------------------------------------------------------

import os
import subprocess


class unittest(object):

	def __init__(self, runner, sUnitName, sTestBinFilename):
		'''
		Constructor.
		'''

		self.m_runner = runner
		self.m_sUnitName = sUnitName
		self.m_sTestBinFilename = sTestBinFilename


	def run(self):
		'''
		Executes the unit test.
		'''

		runner = self.m_runner
		sTestBinFilename = os.environ['TEST_BIN']

		runner.subtest_begin('file_ostream to stdout (host encoding, file-backed)')
		runner.subtest_createfile('OUTPUT_EXPECTED', 'Testing stdout (host encoding)\n')
		iRet = subprocess.call(
			[sTestBinFilename, '-o'],
			stdin = None,
			stdout = open(os.environ['SUBTEST_OUTPUT'], 'wb')
		)
		runner.subtest_end(iRet)

		runner.subtest_begin('file_ostream to file (host encoding)')
		runner.subtest_createfile('OUTPUT_EXPECTED', 'Testing file (host encoding)\n')
		iRet = subprocess.call(
			[sTestBinFilename, '-f', os.environ['SUBTEST_OUTPUT']],
			stdin = None,
			stdout = None
		)
		runner.subtest_end(iRet)

		runner.subtest_begin('file_ostream to file (UTF-8)')
		runner.subtest_createfile('OUTPUT_EXPECTED',
			'\x54\x65\x73\x74\x69\x6e\x67\x20\x66\x69\x6c\x65\x20\x28\x55\x54\x46\x2d\x38\x20\x65' +
			'\x6e\x63\x6f\x64\x69\x6e\x67\x29\x0a'
		)
		iRet = subprocess.call(
			[sTestBinFilename, '-f', os.environ['SUBTEST_OUTPUT'], '-utf8'],
			stdin = None,
			stdout = None
		)
		runner.subtest_end(iRet)

		runner.subtest_begin('file_ostream to file (UTF-16BE)')
		runner.subtest_createfile('OUTPUT_EXPECTED',
			'\x00\x54\x00\x65\x00\x73\x00\x74\x00\x69\x00\x6e\x00\x67\x00\x20\x00\x66\x00\x69' +
			'\x00\x6c\x00\x65\x00\x20\x00\x28\x00\x55\x00\x54\x00\x46\x00\x2d\x00\x31\x00\x36' +
			'\x00\x42\x00\x45\x00\x20\x00\x65\x00\x6e\x00\x63\x00\x6f\x00\x64\x00\x69\x00\x6e' +
			'\x00\x67\x00\x29\x00\x0a'
		)
		iRet = subprocess.call(
			[sTestBinFilename, '-f', os.environ['SUBTEST_OUTPUT'], '-utf16be'],
			stdin = None,
			stdout = None
		)
		runner.subtest_end(iRet)

		runner.subtest_begin('file_ostream to file (UTF-32LE)')
		runner.subtest_createfile('OUTPUT_EXPECTED',
			'\x54\x00\x00\x00\x65\x00\x00\x00\x73\x00\x00\x00\x74\x00\x00\x00\x69\x00\x00\x00' +
			'\x6e\x00\x00\x00\x67\x00\x00\x00\x20\x00\x00\x00\x66\x00\x00\x00\x69\x00\x00\x00' +
			'\x6c\x00\x00\x00\x65\x00\x00\x00\x20\x00\x00\x00\x28\x00\x00\x00\x55\x00\x00\x00' +
			'\x54\x00\x00\x00\x46\x00\x00\x00\x2d\x00\x00\x00\x33\x00\x00\x00\x32\x00\x00\x00' +
			'\x4c\x00\x00\x00\x45\x00\x00\x00\x20\x00\x00\x00\x65\x00\x00\x00\x6e\x00\x00\x00' +
			'\x63\x00\x00\x00\x6f\x00\x00\x00\x64\x00\x00\x00\x69\x00\x00\x00\x6e\x00\x00\x00' +
			'\x67\x00\x00\x00\x29\x00\x00\x00\x0a\x00\x00\x00'
		)
		iRet = subprocess.call(
			[sTestBinFilename, '-f', os.environ['SUBTEST_OUTPUT'], '-utf32le'],
			stdin = None,
			stdout = None
		)
		runner.subtest_end(iRet)

File: src/unittest/0250-file_ostream.cxx
/* -*- coding: utf-8; mode: c++; tab-width: 3 -*-

Copyright 2011, 2012, 2013
Raffaello D. Di Napoli

This file is part of Application-Building Components (henceforth referred to as ABC).

ABC is free software: you can redistribute it and/or modify it under the terms of the GNU General
Public License as published by the Free Software Foundation, either version 3 of the License, or (at
your option) any later version.

ABC is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the
implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public
License for more details.

You should have received a copy of the GNU General Public License along with ABC. If not, see
<http://www.gnu.org/licenses/>.
--------------------------------------------------------------------------------------------------*/

#include <abc/module.hxx>
#include <abc/file_iostream.hxx>
#include <abc/trace.hxx>
using namespace abc;



class test_module :
	public module_impl<test_module> {
public:

	int main(vector<cstring const> const & vsArgs) {
		abc_trace_fn((/*vsArgs*/));

		std::shared_ptr<file_ostream> pfos;
		text::encoding enc(text::encoding::host);
		cstring sName, sEnc(SL("host"));

		size_t cArgs(vsArgs.get_size());
		if (cArgs >= 2 && vsArgs[1] == SL("-o")) {
			pfos = file_ostream::get_stdout();
			sName = SL("stdout");
		} else if (cArgs >= 3 && vsArgs[1] == SL("-f") && vsArgs[2]) {
			if (cArgs == 4 && vsArgs[3] == SL("-utf8")) {
				enc = text::encoding::utf8;
				sEnc = SL("UTF-8");
			} else if (cArgs == 4 && vsArgs[3] == SL("-utf16be")) {
				enc = text::encoding::utf16be;
				sEnc = SL("UTF-16BE");
			} else if (cArgs == 4 && vsArgs[3] == SL("-utf32le")) {
				enc = text::encoding::utf32le;
				sEnc = SL("UTF-32LE");
			}
			pfos = std::make_shared<file_ostream>(vsArgs[2]);
			sName = SL("file");
		}

		if (!pfos) {
			return EXIT_FAILURE;
		}
		pfos->set_encoding(enc);
		*pfos << SL("Testing ")
				<< sName
				<< SL(" (")
				<< sEnc
				<< SL(" encoding)\n");

		// Test results determined by external program.
		return EXIT_SUCCESS;
	}
};

ABC_DECLARE_MODULE_IMPL_CLASS(test_module)

